feat: sdk 1008 manual session

itsdebs opened the pull request:
**Fixes** # (*issue*)

## Type of change
- [ ] Bug fix (non-breaking change which fixes an issue)
- [ ] New feature (non-breaking change which adds functionality)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] This change requires a documentation update
## Checklist:
- [ ] Version upgraded (project, README, gradle, podspec etc)
- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code, particularly in hard-to-understand areas
- [ ] I have added unit tests for the code
- [ ] I have made corresponding changes to the documentation

itsdebs committed:
fix: save auto track session

itsdebs committed:
feat: Removed application out of ConfigurationAndroid.
feat: removed json adapter out of configuration.
feat: changed to lazy configuration initialization to help restoring it from storage.

Signed-off-by: Debanjan Chatterjee <debanjanchatterjee99@gmail.com>

itsdebs committed:
fix: manual session to continue if track auto session is false
if trackAutoSesssion is true, manual session will continue till timeout

Signed-off-by: Debanjan Chatterjee <debanjanchatterjee99@gmail.com>

itsdebs committed:
fix: according to spec changes
session is not dependent on trackLifeCycle
If previous session(manual or automatic) is not timed out,
 no new session will be created

itsdebs committed:
Merge branch 'feat/android' into feature/sdk-1008-manual-session

# Conflicts:
#	android/src/main/java/com/rudderstack/android/RudderAnalytics.kt
#	android/src/main/java/com/rudderstack/android/compat/RudderAnalyticsBuilderCompat.java
#	android/src/test/java/com/rudderstack/android/android/RudderAnalyticsTest.kt
#	core/src/main/java/com/rudderstack/core/Analytics.kt
#	core/src/main/java/com/rudderstack/core/compat/AnalyticsBuilderCompat.java
#	core/src/main/java/com/rudderstack/core/compat/ConfigurationBuilder.java
#	core/src/test/java/com/rudderstack/core/internal/plugins/EventSizeFilterPluginTest.kt
#	core/src/test/java/com/rudderstack/core/internal/plugins/GDPRPluginTest.kt
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/Values.kt

ChryssaAliferi reviewed (changes requested):
Great effort with the analytics test and `TestAnalytics` provider!! But lets not create more `generateTestAnalytics` methods.
In this patch file I have refactored both `generateTestAnalytics` methods to one, the `fun provideAnalytics(...)`:
[feat__refactors_provideAnalytics_test_provider.patch](https://github.com/rudderlabs/rudder-sdk-android/files/15153656/feat__refactors_provideAnalytics_test_provider.patch)


Take a look and let me know! This is way more organised and adjustable on all test/business logic cases.

Also I need us to merge the branch with the current `feat/android` because there are conflicts that make me unsure on what to review. When we do this I will try to review once more!

itsdebs committed:
fix: fix merging. fix test cases

ChryssaAliferi commented on android/src/main/java/com/rudderstack/android/utilities/SessionUtils.kt:
```diff
@@ -111,10 +111,29 @@ internal fun Analytics.initializeSessionManagement(savedSessionId: Long? = null,
+    ((currentConfigurationAndroid?.trackAutoSession != true)
+            && androidStorage.trackAutoSession)
+
+private fun Analytics.updateSessionLastActiveTimestamp() {
```
You can simplify and use directly the userSessionsState value, like that:

```
private fun Analytics.updateSessionLastActiveTimestamp() {
    userSessionState?.value?.let {
        it.copy(lastActiveTimestamp = defaultLastActiveTimestamp)
    }
}
```

ChryssaAliferi commented on core/src/main/java/com/rudderstack/core/internal/ConfigDownloadServiceImpl.kt:
```diff
@@ -31,7 +32,8 @@ import java.util.concurrent.atomic.AtomicReference
     private val writeKey: String,
-    webService: WebService? = null,
+    private val jsonAdapter: JsonAdapter,
+    webService: WebService? = null
```
`private val webService: WebService? = null`

ChryssaAliferi commented on samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt:
```diff
@@ -22,34 +22,43 @@ object RudderAnalyticsUtils {
+//                dataPlaneUrl = DATA_PLANE_URL,
+//                controlPlaneUrl = CONTROL_PLANE_URL,
+//                recordScreenViews = true,
+                copy(
```
Please would you like to explain more about this `copy` method and that callback added here? Frankly, it does not look much user-friendly.

itsdebs commented on core/src/test/java/com/rudderstack/core/internal/plugins/EventSizeFilterPluginTest.kt:
```diff
@@ -9,13 +10,19 @@ import com.rudderstack.core.internal.CentralPluginChain
 class EventSizeFilterPluginTest {
-
+    @get:Rule
+    val mockkRule = MockKRule(this)
```
This came up due to merging, EventSizeFiltering is a separate feature and was implemented in an old PR
[pull-request](https://github.com/rudderlabs/rudder-sdk-android/pull/396)

itsdebs replied:
We can remove these in separate PR

itsdebs commented on core/src/main/java/com/rudderstack/core/internal/plugins/EventSizeFilterPlugin.kt:
```diff
@@ -14,17 +16,24 @@ import java.util.concurrent.atomic.AtomicReference
 class EventSizeFilterPlugin : Plugin {
 
     private val currentConfigurationAtomic = AtomicReference<Configuration?>()
+    private var _analyticsRef = WeakReference<Analytics>(null)
```
I have not been through this plugin, this plugin was part of the event filtering task, can we deal with it in a separate PR? 
You can create a linear ticket for improvements.

itsdebs committed:
Merge branch 'feat/android' into feature/sdk-1008-manual-session

# Conflicts:
#	android/src/main/java/com/rudderstack/android/RudderAnalytics.kt
#	android/src/main/java/com/rudderstack/android/storage/RudderEntityFactory.kt
#	android/src/test/java/com/rudderstack/android/internal/plugins/ReinstatePluginTest.kt
#	core/src/main/java/com/rudderstack/core/internal/AnalyticsDelegate.kt
#	core/src/main/java/com/rudderstack/core/internal/DataUploadServiceImpl.kt
#	core/src/test/java/com/rudderstack/core/internal/ConfigDownloadServiceImplTest.kt
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt

itsdebs committed:
feat: merge commit

itsdebs committed:
feat: change applyConfig signature

itsdebs committed:
fix: rudder analytics utils config initializer

itsdebs committed:
Merge branch 'feat/android' into feature/sdk-1008-manual-session

# Conflicts:
#	android/src/main/java/com/rudderstack/android/ConfigurationAndroid.kt
#	android/src/main/java/com/rudderstack/android/RudderAnalytics.kt
#	android/src/main/java/com/rudderstack/android/compat/ConfigurationAndroidBuilder.java
#	android/src/main/java/com/rudderstack/android/compat/RudderAnalyticsBuilderCompat.java
#	android/src/main/java/com/rudderstack/android/internal/plugins/PlatformInputsPlugin.kt
#	android/src/main/java/com/rudderstack/android/utilities/SessionUtils.kt
#	android/src/test/java/com/rudderstack/android/RudderAnalyticsTest.kt
#	android/src/test/java/com/rudderstack/android/internal/infrastructure/ReinstatePluginTest.kt
#	android/src/test/java/com/rudderstack/android/utilities/SessionUtilsTest.kt
#	core/src/main/java/com/rudderstack/core/Analytics.kt
#	core/src/main/java/com/rudderstack/core/Configuration.kt
#	core/src/main/java/com/rudderstack/core/compat/ConfigurationBuilder.java
#	core/src/main/java/com/rudderstack/core/internal/DataUploadServiceImpl.kt
#	core/src/test/java/com/rudderstack/core/AnalyticsTest.kt
#	core/src/test/java/com/rudderstack/core/compat/ConfigurationBuilderTest.java
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/mainview/MainViewModel.kt

itsdebs committed:
Merge branch 'feat/android' into feature/sdk-1008-manual-session

# Conflicts:
#	android/src/main/java/com/rudderstack/android/ConfigurationAndroid.kt
#	android/src/main/java/com/rudderstack/android/compat/ConfigurationAndroidBuilder.java
#	android/src/main/java/com/rudderstack/android/storage/AndroidStorageImpl.kt
#	android/src/test/java/com/rudderstack/android/AndroidStorageTest.kt
#	android/src/test/java/com/rudderstack/android/RudderAnalyticsTest.kt
#	android/src/test/java/com/rudderstack/android/internal/infrastructure/ReinstatePluginTest.kt
#	android/src/test/java/com/rudderstack/android/plugins/PlatformInputsPluginTest.kt
#	android/src/test/java/com/rudderstack/android/utilities/SessionUtilsTest.kt
#	core/src/main/java/com/rudderstack/core/internal/AnalyticsDelegate.kt
#	core/src/test/java/com/rudderstack/core/compat/ConfigurationBuilderTest.java
#	samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt

itsdebs committed:
chore: merge changes
chore: fix tests

itsdebs committed:
fix: issue with configuration scope

itsdebs committed:
fix: change session id to long instead of string

1abhishekpandey commented on samples/sample-kotlin-android/src/main/kotlin/com/rudderstack/android/sampleapp/analytics/RudderAnalyticsUtils.kt:
```diff
@@ -29,20 +29,21 @@ object RudderAnalyticsUtils {
-                logLevel = RudderLogger.LogLevel.DEBUG,
-                autoCollectAdvertId = true,
-            )
+            configurationScope = {
+                dataPlaneUrl = DATA_PLANE_URL
+                controlPlaneUrl = CONTROL_PLANE_URL
+                trackLifecycleEvents = true
+                recordScreenViews = true
+                isPeriodicFlushEnabled = true
+                autoCollectAdvertisingId = true
+                trackAutoSession = true
+                logLevel = RudderLogger.LogLevel.DEBUG
+            }
```
Can you please revert this change for now?

I tried to disable the `lifecycle` flag to `false` and after that also I can see the lifecycle events in our dashboard.

itsdebs committed:
fix: trackLifecycle field fix

itsdebs committed:
fix: dependency on trackLifecycle removed.

itsdebs committed:
feat: Session won't expire in a single run.
feat: reset will now refresh the session id if it exists
fix: manual session continuation

ChryssaAliferi commented:
Closing this PR as all the progress has been included in the new [Kotlin SDK](https://github.com/rudderlabs/rudder-sdk-kotlin).
